// @zappar/zappar-react-three-fiber/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserCompatibility = exports.Loader = exports.ZapparCanvas = exports.ZapparCamera = exports.InstantTracker = exports.ImageTracker = exports.HeadMaskMesh = exports.FaceTracker = exports.FaceLandmark = exports.FaceBufferGeometry = exports.setLogLevel = exports.LogLevel = exports.Pipeline = exports.browserIncompatible = void 0;
const version_1 = __importDefault(require("./version"));
// eslint-disable-next-line no-console
console.log(`Zappar for React Three v${version_1.default}`);
var zappar_threejs_1 = require("@zappar/zappar-threejs");
Object.defineProperty(exports, "browserIncompatible", { enumerable: true, get: function () { return zappar_threejs_1.browserIncompatible; } });
Object.defineProperty(exports, "Pipeline", { enumerable: true, get: function () { return zappar_threejs_1.Pipeline; } });
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return zappar_threejs_1.LogLevel; } });
Object.defineProperty(exports, "setLogLevel", { enumerable: true, get: function () { return zappar_threejs_1.setLogLevel; } });
var FaceBufferGeometry_1 = require("./Components/FaceBufferGeometry");
Object.defineProperty(exports, "FaceBufferGeometry", { enumerable: true, get: function () { return __importDefault(FaceBufferGeometry_1).default; } });
var FaceLandmarkGroup_1 = require("./Components/trackers/FaceLandmarkGroup");
Object.defineProperty(exports, "FaceLandmark", { enumerable: true, get: function () { return __importDefault(FaceLandmarkGroup_1).default; } });
var FaceTrackerGroup_1 = require("./Components/trackers/FaceTrackerGroup");
Object.defineProperty(exports, "FaceTracker", { enumerable: true, get: function () { return __importDefault(FaceTrackerGroup_1).default; } });
var HeadMaskMesh_1 = require("./Components/HeadMaskMesh");
Object.defineProperty(exports, "HeadMaskMesh", { enumerable: true, get: function () { return __importDefault(HeadMaskMesh_1).default; } });
var ImageTrackerGroup_1 = require("./Components/trackers/ImageTrackerGroup");
Object.defineProperty(exports, "ImageTracker", { enumerable: true, get: function () { return __importDefault(ImageTrackerGroup_1).default; } });
var InstantTrackerGroup_1 = require("./Components/trackers/InstantTrackerGroup");
Object.defineProperty(exports, "InstantTracker", { enumerable: true, get: function () { return __importDefault(InstantTrackerGroup_1).default; } });
var ZapparCamera_1 = require("./Components/ZapparCamera");
Object.defineProperty(exports, "ZapparCamera", { enumerable: true, get: function () { return __importDefault(ZapparCamera_1).default; } });
var ZapparCanvas_1 = require("./Components/ZapparCanvas");
Object.defineProperty(exports, "ZapparCanvas", { enumerable: true, get: function () { return __importDefault(ZapparCanvas_1).default; } });
var Loader_1 = require("./Components/util/Loader");
Object.defineProperty(exports, "Loader", { enumerable: true, get: function () { return __importDefault(Loader_1).default; } });
var Compatibility_1 = require("./Components/util/Compatibility");
Object.defineProperty(exports, "BrowserCompatibility", { enumerable: true, get: function () { return __importDefault(Compatibility_1).default; } });
// export { default as CameraEnvironmentMap } from "./Components/CameraEnvironmentMap";
