// @zappar/zappar-cv/lib/worker-client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.launchWorker = exports.messageManager = void 0;
const messages_1 = require("./messages");
exports.messageManager = new messages_1.MsgManager();
let Worker;
if (typeof Z_STANDALONE !== 'undefined' && Z_STANDALONE) {
    Worker = require('worker-loader?inline=fallback!./worker');
}
else {
    Worker = require('worker-loader!./worker');
}
if (Worker.default)
    Worker = Worker.default;
function launchWorker() {
    return __awaiter(this, void 0, void 0, function* () {
        let worker = new Worker();
        worker.postMessage({
            t: "wasm",
            url: require("./zcv.wasm").default
        });
        yield waitForLoad(worker);
        function sendOutgoing() {
            let msgs = exports.messageManager.getOutgoingMessages();
            for (let msg of msgs) {
                worker.postMessage(msg.msg, msg.transferables);
            }
        }
        exports.messageManager.onOutgoingMessage.bind(sendOutgoing);
        sendOutgoing();
        worker.addEventListener("message", evt => {
            exports.messageManager.postIncomingMessage(evt.data);
        });
    });
}
exports.launchWorker = launchWorker;
function waitForLoad(w) {
    return new Promise(resolve => {
        let listener = (msg) => {
            if (msg.data === "loaded") {
                w.removeEventListener("message", listener);
                resolve();
            }
        };
        w.addEventListener("message", listener);
    });
}
